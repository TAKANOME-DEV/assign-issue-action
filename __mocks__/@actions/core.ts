// import {InputOptions} from '@actions/core';

export const getInput = jest.fn().mockImplementation((name: string) => name);
// jest.spyOn(core, 'getInput').mockImplementation((name: string) => {
//   return inputs[name]
// })
// export const getInput = jest
//   .spyOn(core, 'getInput')
//   .mockImplementation((name, options) => {
//     return jest.requireActual('@actions/core').getInput(name, options);
//   });
