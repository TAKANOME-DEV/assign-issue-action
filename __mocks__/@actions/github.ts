export const context = {
  payload: {
    comment: {
      user: {
        login: 'John',
      },
      body: '/assign-to-me',
      html_url: 'https://github.com/TAKANOME-DEV/testing',
    },
    issue: {
      labels: [],
      assignee: null,
      number: 1,
    },
  },
  repo: {
    owner: 'TAKANOME-DEV',
    repo: 'testing',
  },
  action: 'created',
};

const mockApi = {
  rest: {
    issues: {
      addAssignees: jest.fn(),
      addLabels: jest.fn(),
      createComment: jest.fn(),
      removeLabel: jest.fn(),
      removeAssignee: jest.fn(),
    },
    search: {
      issuesAndPullRequests: jest.fn().mockImplementation(() => ({
        status: 200,
        body: {
          items: [
            {
              number: 1,
              labels: [],
              assignee: {
                login: 'John',
              },
            },
          ],
        },
      })),
    },
  },
};

export const getOctokit = jest.fn().mockImplementation(() => mockApi);
