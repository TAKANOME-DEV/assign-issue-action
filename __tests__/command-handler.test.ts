/* eslint-disable filenames/match-regex */

import * as core from '@actions/core';
import * as github from '@actions/github';
import fs from 'fs';
import jsYaml from 'js-yaml';
import path from 'path';
import commandHandler from '../src/lib/command-handler';

/**
 * Helper that reads the `action.yml` and includes the default values
 * for each input as an environment variable, like the Actions runtime does.
 */
function getDefaultValues() {
  // eslint-disable-next-line no-undef
  const yaml = fs.readFileSync(path.join(__dirname, '../action.yml'), 'utf8');
  const { inputs } = jsYaml.load(yaml) as any;

  return Object.keys(inputs).reduce((acc, key) => {
    if ('default' in inputs[key]) {
      return {
        ...acc,
        [`INPUT_${key.toUpperCase()}`]: inputs[key].default,
      };
    } else {
      return acc;
    }
  }, {});
}

console.log('default Values', getDefaultValues());

Object.assign(
  process.env,
  {
    GITHUB_ACTION: 'my-action',
    GITHUB_ACTOR: 'TAKANOME-DEV',
    GITHUB_EVENT_NAME: 'issue_comment',
    GITHUB_EVENT_PATH: path.join(__dirname, '../__mocks__/@actions/github.ts'),
    GITHUB_REF: 'main',
    GITHUB_REPOSITORY: 'TAKANOME-DEV/testing',
    GITHUB_SHA: '123abc',
    GITHUB_TOKEN: '_',
    GITHUB_WORKFLOW: 'my-workflow',
    GITHUB_WORKSPACE: path.join(__dirname, '../__mocks__/@actions/github.ts'),
    HOME: '?',
  },
  getDefaultValues()
);

jest.mock('@actions/core');
jest.mock('@actions/github');

const gh = github.getOctokit('_');
const addAssigneesMock = jest.spyOn(gh.rest.issues, 'addAssignees');
const addLabelsMock = jest.spyOn(gh.rest.issues, 'addLabels');
// const createCommentMock = jest.spyOn(gh.rest.issues, 'createComment');

afterAll(() => jest.restoreAllMocks());

describe('command-handler', () => {
  it('should assigns the user to the issue', async () => {
    await commandHandler(core, github);

    expect(addAssigneesMock).toHaveBeenCalledTimes(1);
    // expect(addAssigneesMock).toHaveBeenCalledWith({
    //   owner: 'takanome_dev',
    //   repo: 'testing',
    //   issue_number: 1,
    //   assignees: ['John'],
    // });
    expect(addLabelsMock).toHaveBeenCalledTimes(1);
    // expect(addLabelsMock).toHaveBeenCalledWith({
    //   owner: 'takanome_dev',
    //   repo: 'testing',
    //   issue_number: 1,
    //   labels: ['Is-Assigned'],
    // });
    // expect(createCommentMock).toHaveBeenCalledTimes(1);
    // expect(createCommentMock).toHaveBeenCalledWith({
    //   totalDays: 7,
    //   body: `This issue [has been assigned](https://github.com/testing) to @John!
    //   It will become unassigned if it isn't closed within 7 days.
    //   A maintainer can also add the **pinned** label to prevent it from being unassigned.`,
    // });

    // const [assignReq, labelReq, commentReq] = requests;
    // console.log({ assignReq, labelReq, commentReq });

    // expect(assignReq.reqBody.assignees).toEqual([
    //   github.context.payload.comment?.user.login,
    // ]);
    // expect(labelReq.reqBody.labels).toEqual(['Is-Assigned']);
    // expect(commentReq.reqBody.body).toMatchSnapshot();
  });
});
