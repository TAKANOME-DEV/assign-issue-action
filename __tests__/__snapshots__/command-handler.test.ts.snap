// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command-handler should assigns the user to the issue 1`] = `
Object {
  "body": "/assign-to-me",
  "html_url": "https://github.com/TAKANOME-DEV/testing",
  "user": Object {
    "login": "John",
  },
}
`;
